[PR] Evidence story filtering

philcable opened the pull request:
@jeremyfelt #review? No merge yet as the design is just something I threw together and likely to change.

Commit:
Remove some body classes from The Evidence style guide page

Commit:
Add a few more stories to the style guide page

Commit:
Update content in top section and add markup for filtering function

Commit:
Add initial scroll handling JavaScript for the filtering feature

Commit:
Add styles for the fixed filter drop-down

Commit:
Add filter markup to The Evidence template

Commit:
Remove `, Academic` from the term name output

Commit:
Localize evidence.js with page title, url, ajax url and nonce data

Commit:
Update `$story_query` args and add a container with data attributes

The data attributes will be leveraged in the JavaScript for an
infinite scroll effect

Commit:
Update JavaScript with history state and AJAX call functions

This commit also refactors how the functions are defined and makes
other syntactical changes throughout for consistency. Apologies in
advance for the confusing diff.

Commit:
Add handlers for the AJAX requests

Commit:
Update styles for the story container and loading indicator

Commit:
Add rewrite rules and a `category` query var for category views on The Evidence page

Commit:
Capture the `category` query var in `template-evidence`

This commit also outputs the corresponding heading, adds `selected`
to the corresponding filter option, and adds tax_query parameters
to the stories query.

Commit:
Include a trailing slash when the url is change by the History API

Commit:
Update `style-guide/js/evidence.js`

Evidently I neglected to run grunt after refactoring the function
definitions.

Commit:
Remove `, Academic` from category headings

Commit:
Include selection value and name as a state object in the pushState method

Commit:
Fix markup so that the selection name doesn't include any weird spacing

Commit:
Add a popstate handler to update the page when browser nav buttons are used

There's almost certainly a more elegant way to accomplish this,
but this should do the trick for now.

Commit:
Replace the history state when the page is first visited

This will ensure that navigating back to it via the browser back
button will yield the expected results.

Commit:
Add a class to the intro section and adjust markup around the filter drop-down

Commit:
Update selectors to accommodate the new markup

Commit:
Add styles for the filter

Just some initial ideas; subject to change.

Commit:
Make adjustments to meet coding standards

jeremyfelt commented on js/evidence.js:
```diff
@@ -1,26 +1,196 @@
+	 * Scroll to the filter wrapper when an option is selected.
+	 */
+	function scroll_to_filter() {
+		var filters_top = Math.round($filters.offset().top - offset);
```
It may be possible to capture this when the script first initializes. Though, if this isn't hit that often, it may not matter much.

jeremyfelt commented on template-evidence.php:
```diff
@@ -18,23 +33,56 @@
 				<?php
 				$stories_query_args = array(
 					'post_type' => 'drive_story',
-					'posts_per_page' => -1,
```
We should probably still keep an explicit default of 10 or something here so that the "posts per page" setting in WP doesn't accidentally affect it.

jeremyfelt commented on functions.php:
```diff
@@ -408,6 +418,102 @@ public function ajax_comments() {
+		$stories_query = new WP_Query( $stories_query_args );
+
+		if ( $stories_query->have_posts() ) {
+			$stories = array();
```
I would move this outside of the `if` conditional so that in the super crazy scenario that there are no posts, the `wp_json_encode()` at the end of this method has an empty array to work with.

jeremyfelt commented on js/evidence.js:
```diff
@@ -1,26 +1,196 @@
+		var state = { name: name, value: value },
+			url = value ? evidence.default_url + 'category/' + value + '/' : evidence.default_url;
+
+		history.pushState(state, null, url);
```
The second argument in `history.pushState()` should be a string. I'd go with an empty string rather than `null`.

jeremyfelt commented on js/evidence.js:
```diff
@@ -1,26 +1,196 @@
+			name = value ? option.text() : '&nbsp;',
+			state = { name: name, value: value };
+
+		history.replaceState(state, null, null);
```
Same as with `pushState()`, the title parameter can be `''`. I _think_ we can leave off the 3rd parameter (url) if one isn't available so that it uses the current URL.

jeremyfelt commented on js/evidence.js:
```diff
@@ -1,26 +1,196 @@
-	//'use strict';
+	'use strict';
+
+	var $filters = $('#story-filters'),
```
We haven't talked about this at all... :grin: ... but I'd like to start pushing for using jQuery's JS style guide - https://contribute.jquery.org/style-guide/js/. I **wouldn't worry** about changing all of this code, as it's a project in progress, but we should talk about it for the future. The biggest adjustment is the use of `"` rather than `'` and more liberal spacing, similar to WordPress and PHP.

jeremyfelt commented on js/evidence.js:
```diff
@@ -1,26 +1,196 @@
+		fetch_and_display_stories(data, 'filtered');
+	};
+
+}(jQuery, window, document, evidence));
```
I get `evidence is not defined` in the style guide because `evidence` is not defined. :) Not really a big deal as replacing all of the ajax functionality that WP provides would be kind of ridiculous. The file could start with `var evidence = evidence || {};` as the first line so that it is available no matter what. There may still be other problems, but it would help avoid the first error.

jeremyfelt reviewed (changes requested):
Looking great! I left some comments inline. A few of them should probably be addressed beforehand, others aren't so important.

I think we briefly talked about using the REST API for this, but then there was some wonder what it would look like to have a public REST API post type that isn't public in other ways. We should figure that out for the future. It may also make it easier to pull content in for the style guide. :)

Commit:
Add a `posts_per_page` argument to the story queries

Commit:
Move the `$stories` array outside the `if` conditional

Commit:
Update `history.pushState` and `history.replaceState` parameters

Commit:
Rename cached selectors

This should (hopefully) make the code a bit more intuitive, as
these names better match the selectors they represent (or at least
their intent). It also consolidates the `filters_top` and
`$wrapper_offset` variables, as they served much the same purpose.
These changes will allow for removing an extra div in the template.

Commit:
Remove `#story-filters` div as it is no longer needed

Additionally, this adjusts how the padding is handled for the
`.intro` section.

Commit:
Update styles to accommodate padding changes

Commit:
Recalculate `filter_top` value on resize

Commit:
Update the Evidence style guide page (as much as possible, anyway)

Commit:
Build appropriate titles for the Evidence page

Commit:
Add REST API support for the `drive_story` content type

Commit:
Include mobile image data in the REST API response for `drive_story` posts

Commit:
Refactor AJAX call to use the REST API

Commit:
Remove AJAX request handlers

Commit:
Update `evidence` variable on the Evidence style guide

philcable commented:
These next five commits refactor things to use the REST API. It might be a little rough...

philcable commented:
@jeremyfelt #review?

jeremyfelt committed:
Update the style guide JavaScript

jeremyfelt commented:
Looks great @philcable. I'm impressed by how smoothly the REST API fit. The only small issue I saw is a JS error if a mobile image is not assigned to a story. I'm not sure if that's worth worrying about now as it doesn't break functionality and we should have mobile images assigned to all stories.
